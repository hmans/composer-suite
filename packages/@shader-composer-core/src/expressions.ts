import { glslRepresentation } from "./glslRepresentation"

const zip = (a: TemplateStringsArray, b: any[]) => a.map((k, i) => [k, b[i]])

export type Expression = {
  _: "Expression"
  values: any[]
  render: () => string
  toString: () => string
}

export const glsl = (strings: TemplateStringsArray, ...values: any[]): Expression => {
  const render = () =>
    zip(
      strings,
      values.map((v) => glslRepresentation(v))
    )
      .flat()
      .join("")

  return {
    _: "Expression",
    values,
    render,
    toString: render
  }
}

/** A shortcut for the `glsl` tagged template literal helper. */
export const $ = glsl

export function isExpression(v: any): v is Expression {
  return v && v._ === "Expression"
}
