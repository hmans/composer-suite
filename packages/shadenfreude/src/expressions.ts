import { glslRepresentation } from "./glslRepresentation"

export type Expression = {
  _: "Expression"
  values: any[]
  render: () => string
}

const zip = (a: TemplateStringsArray, b: any[]) => a.map((k, i) => [k, b[i]])

export const code = (
  strings: TemplateStringsArray,
  ...values: any[]
): Expression => ({
  _: "Expression",

  values,

  render: () =>
    zip(strings, values.map(glslRepresentation))
      .flat()
      .join("")
})

export function isExpression(v: any): v is Expression {
  return v && v._ === "Expression"
}
