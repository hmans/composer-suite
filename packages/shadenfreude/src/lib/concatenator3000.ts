export type Parts = any[]

const compact = (p: any) => !!p

const indent = (p: string) => "  " + p

export const block = (...parts: Parts): Parts => {
  const flattened = flatten(parts)
  return flattened.length > 0 ? ["{", flattened.map(indent), "}"] : []
}

export const concatenate = (...parts: Parts) => flatten(...parts).join("\n")

export const line = (...parts: Parts) => flatten(...parts).join(" ")

export const flatten = (...parts: Parts): Parts =>
  parts
    .filter(compact)
    .map((p) => (Array.isArray(p) ? flatten(...p) : p))
    .flat()

export const comment = (...parts: Parts) => line("/*", ...parts, "*/")

export const statement = (...parts: Parts) => line(...parts) + ";"

export const assignment = (left: string, right: string) =>
  statement(left, "=", right)

export const identifier = (...parts: Parts) =>
  parts
    .flat()
    .filter(compact)
    .join("_")
    .replace(/_{2,}/g, "_")

export const sluggify = (s: string) =>
  s.replace(/[^a-zA-Z0-9]/g, "_").replace(/_{2,}/g, "_")
