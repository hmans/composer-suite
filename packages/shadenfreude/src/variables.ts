import { glslType } from "./helpers"
import {
  GLSLType,
  isShaderNode,
  isVariable,
  ShaderNode,
  Value,
  Variable
} from "./types"

export const variablesToNodes = new Map<Variable, ShaderNode>()

export function variable<T extends GLSLType>(
  type: T,
  value?: Value<T>
): Variable<T> {
  /* By default, we'll give the variable a random name. */
  const name = `var_${type}_${Math.floor(Math.random() * 100000)}`

  return {
    _variable: true,
    name,
    type,
    value
  }
}

export const float = (value?: Value<"float">) => variable("float", value)

export const bool = (value?: Value<"bool">) => variable("bool", value)

export const vec2 = (value?: Value<"vec2">) => variable("vec2", value)

export const vec3 = (value?: Value<"vec3">) => variable("vec3", value)

export const vec4 = (value?: Value<"vec4">) => variable("vec4", value)

export const mat3 = (value?: Value<"mat3">) => variable("mat3", value)

export const mat4 = (value?: Value<"mat4">) => variable("mat4", value)

/**
 * Creates a new variable with the GLSL type inferred from the specified value.
 */
export function inferVariable(a: Value): Variable {
  if (isVariable(a)) {
    return variable(a.type, a)
  } else if (isShaderNode(a)) {
    if (a.outputs.value === undefined) {
      throw new Error("No value output on node")
    } else {
      return inferVariable(a.outputs.value)
    }
  } else {
    return variable(glslType(a), a)
  }
}
