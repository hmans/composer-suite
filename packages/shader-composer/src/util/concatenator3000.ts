export type Part = any

export const isPresent = (p: Part) => !!p

export const concatenate = (...parts: Part[]): string =>
	parts
		.flat(Infinity)
		.filter(isPresent)
		.join("\n")

export const identifier = (...parts: Part[]): string =>
	parts
		.filter(isPresent)
		.join("_")
		.replace(/_{2,}/g, "_")

export const statement = (...parts: Part[]): string =>
	parts.filter(isPresent).join(" ") + ";"

export const assignment = (...parts: Part[]): string => statement(parts.join(" = "))

export const block = (...parts: Part[]): Part[] => [
	"{",
	concatenate(parts)
		.split("\n")
		.map((line) => "  " + line)
		.join("\n"),
	"}"
]

export const sluggify = (s: string) => s.replace(/[^a-zA-Z0-9]/g, "_")
