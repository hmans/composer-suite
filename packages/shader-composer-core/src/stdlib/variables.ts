import { $, Expression } from "../expressions"
import { frame } from "../ticker"
import { GLSLType, injectAPI, JSTypes, Unit, UnitConfig } from "../units"
import { $localToViewSpace, $localToWorldSpace } from "./spaces"
import { Bool, Int, Mat4, Vec2, Vec3 } from "./values"

/**
 * Returns the current fragment's on-screen coordinate.
 */
export const FragmentCoordinate = Vec2($`gl_FragCoord.xy`, {
  name: "Fragment Coordinate",
  only: "fragment"
})

/**
 * In instanced rendering, will return the instance ID.
 * Wraps the `gl_InstanceID` GLSL built-in.
 */
export const InstanceID = Int($`gl_InstanceID`, {
  name: "Instance ID",
  varying: "flat"
})

/**
 * Returns the verte ID.
 * Wraps the `gl_VertexID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const VertexID = Int($`gl_VertexID`, {
  name: "Vertex ID",
  only: "vertex"
})

export const CameraPosition = Vec3($`cameraPosition`, {
  name: "Camera Position"
})

export const ViewMatrix = Mat4($`viewMatrix`, {
  name: "View Matrix"
})

export const ModelMatrix = Mat4($`modelMatrix`, {
  name: "Model Matrix"
})

export const ModelViewMatrix = Mat4($`modelViewMatrix`, {
  name: "ModelView Matrix"
})

export const NormalMatrix = Mat4($`normalMatrix`, {
  name: "Normal Matrix"
})

export const ProjectionMatrix = Mat4($`projectionMatrix`, {
  name: "Projection Matrix"
})

/**
 * Returns true if instanced rendering is enabled, false if it is not.
 */
export const UsingInstancing = Bool($`
  #ifdef USE_INSTANCING
    true
  #else
    false
  #endif
`)

/**
 * Returns the instance matrix. Please note that this is only available when
 * instanced rendering is enabled.
 */
export const InstanceMatrix = Mat4($`instanceMatrix`, {
  name: "Instance Matrix",
  only: "vertex"
})

export const UV = Vec2($`uv`, {
  name: "UV",
  varying: true
})

const Vec3WithSpaceConversions = (expr: Expression, name: string) =>
  injectAPI(Vec3(expr, { name, varying: true }), (unit) => ({
    get world() {
      return Vec3($localToWorldSpace(unit), {
        varying: true,
        name: `${name} (World Space)`
      })
    },

    get view() {
      return Vec3($localToViewSpace(unit), {
        varying: true,
        name: `${name} (View Space)`
      })
    }
  }))

export const VertexPosition = Vec3WithSpaceConversions(
  $`position`,
  "Vertex Position"
)
export const VertexNormal = Vec3WithSpaceConversions($`normal`, "Vertex Normal")

export const ViewDirection = Vec3(
  $`vec3(-${ViewMatrix}[0][2], -${ViewMatrix}[1][2], -${ViewMatrix}[2][2])`,
  { varying: true, name: "View Direction" }
)

export const IsFrontFacing = Bool($`gl_FrontFacing`, { only: "fragment" })

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })

export type UniformUnit<
  T extends GLSLType,
  J extends JSTypes[T] = JSTypes[T]
> = Unit<T> & {
  value: J
}

export const UniformUnit = <T extends GLSLType, J extends JSTypes[T]>(
  type: T,
  initialValue: J,
  extras?: Partial<UnitConfig<T>>
): UniformUnit<T, J> => {
  const uniform = { value: initialValue }

  /* Create the actual unit that represents the uniform. */
  const unit = Unit(type, undefined, {
    name: `Uniform (${type})`,
    ...extras,
    uniform
  })

  /* Return the unit with some API bits mixed in. */
  return injectAPI(unit, () => ({
    set value(v: J) {
      uniform.value = v
    },

    get value(): J {
      return uniform.value
    }
  }))
}

/**
 * Provides a uniform unit holding a representation of time. The time value
 * stored is not an absolute time, so multiple instances of this unit will not
 * be synchronized. If you require synchronization, please either reuse the
 * same instance of this unit, or use `GlobalTime` instead.
 *
 * @param initial The initial time value to start with. (Default: 0)
 */
export const Time = (initial: number = 0) => {
  const uniform = UniformUnit("float", initial, {
    name: "Time Uniform",

    update: (dt) => {
      uniform.value += dt
    }
  })

  return uniform
}

/**
 * A global time uniform unit that can be safely used across multiple shaders,
 * wherever synchronization is required,
 * and as a default value for `time` inputs of other unit implementations, to prevent
 * shaders from being spammed by multiple uniforms all holding different
 * representations of time.
 */
export const GlobalTime = Time()

/**
 * An absolute time value, in seconds, that remains stable during the duration
 * of the currently rendered frame. This is useful for creating effects that
 * require accurate synchronization across multiple shaders.
 */
export const FrameTime = UniformUnit("float", frame.time, {
  name: "Frame Time",

  update: () => {
    FrameTime.value = frame.time
  }
})

/**
 * An absolute count of the number of frames that have been rendered so far.
 * This is useful for creating effects that require accurate synchronization
 * across multiple shaders, and prefer to use an integer frame count instead
 * of a floating point time value like {@link FrameTime} provides.
 */
export const FrameCount = UniformUnit("int", frame.count, {
  name: "Frame Count",

  update: () => {
    FrameCount.value = frame.count
  }
})

export const Resolution = UniformUnit(
  "vec2",
  { x: 0, y: 0 },
  {
    name: "Current Render Resolution",

    update: () => {
      "Please customize the Resolution unit with the relevant code for your environment."
    }
  }
)

export const CameraNear = UniformUnit("float", 0 as number, {
  name: "Camera Near Plane",

  update: () => {
    console.warn(
      "Please customize the CameraNear unit with the relevant code for your environment."
    )
  }
})

export const CameraFar = UniformUnit("float", 0 as number, {
  name: "Camera Far Plane",

  update: () => {
    console.warn(
      "Please customize the CameraNear unit with the relevant code for your environment."
    )
  }
})

export const ScreenUV = Vec2($`${FragmentCoordinate} / ${Resolution}`, {
  name: "Screen UV"
})
