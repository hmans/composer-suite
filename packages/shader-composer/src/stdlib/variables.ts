import { PerspectiveCamera, Vector2 } from "three"
import { $, Expression } from "../expressions"
import { GLSLType, injectAPI, Input, JSTypes, Unit, UnitConfig } from "../units"
import { ScaleAndOffset } from "./math"
import { $localToViewSpace, $localToWorldSpace } from "./spaces"
import { Bool, Int, Mat4, Vec2, Vec3 } from "./values"

/**
 * Returns the current fragment's on-screen coordinate.
 */
export const FragmentCoordinate = Vec2($`gl_FragCoord.xy`, {
  name: "Fragment Coordinate",
  only: "fragment"
})

/**
 * In instanced rendering, will return the instance ID.
 * Wraps the `gl_InstanceID` GLSL built-in.
 */
export const InstanceID = Int(0, {
  name: "Instance ID",
  varying: "flat"
})

/**
 * Returns the verte ID.
 * Wraps the `gl_VertexID` GLSL built-in.
 *
 * Note: available in vertex shader only!
 */
export const VertexID = Int($`gl_VertexID`, {
  name: "Vertex ID",
  only: "vertex"
})

export const CameraPosition = Vec3($`cameraPosition`, {
  name: "Camera Position"
})

export const ViewMatrix = Mat4($`viewMatrix`, {
  name: "View Matrix"
})

export const ModelMatrix = Mat4($`modelMatrix`, {
  name: "Model Matrix"
})

export const ModelViewMatrix = Mat4($`modelViewMatrix`, {
  name: "ModelView Matrix"
})

export const NormalMatrix = Mat4($`normalMatrix`, {
  name: "Normal Matrix"
})

export const ProjectionMatrix = Mat4($`projectionMatrix`, {
  name: "Projection Matrix"
})

/**
 * Returns true if instanced rendering is enabled, false if it is not.
 */
export const UsingInstancing = Bool($`
  #ifdef USE_INSTANCING
    true
  #else
    false
  #endif
`)

/**
 * Returns the instance matrix. Please note that this is only available when
 * instanced rendering is enabled.
 */
export const InstanceMatrix = Mat4($`instanceMatrix`, {
  name: "Instance Matrix",
  only: "vertex"
})

export const UV = Vec2($`uv`, {
  name: "UV",
  varying: true
})

const Vec3WithSpaceConversions = (expr: Expression, name: string) =>
  injectAPI(Vec3(expr, { name, varying: true }), (unit) => ({
    get world() {
      return Vec3($localToWorldSpace(unit), {
        varying: true,
        name: `${name} (World Space)`
      })
    },

    get view() {
      return Vec3($localToViewSpace(unit), {
        varying: true,
        name: `${name} (View Space)`
      })
    }
  }))

export const VertexPosition = Vec3WithSpaceConversions(
  $`position`,
  "Vertex Position"
)
export const VertexNormal = Vec3WithSpaceConversions($`normal`, "Vertex Normal")

export const ViewDirection = Vec3(
  $`vec3(-${ViewMatrix}[0][2], -${ViewMatrix}[1][2], -${ViewMatrix}[2][2])`,
  { varying: true, name: "View Direction" }
)

export const IsFrontFacing = Bool($`gl_FrontFacing`, { only: "fragment" })

export const Attribute = <T extends GLSLType>(type: T, name: string) =>
  Unit(type, $`${name}`, {
    name: `Attribute: ${name}`,
    varying: true,
    vertex: {
      header: $`attribute ${type} ${name};`
    }
  })

export type UniformUnit<
  T extends GLSLType,
  J extends JSTypes[T] = JSTypes[T]
> = Unit<T> & {
  value: J
}

export const UniformUnit = <T extends GLSLType, J extends JSTypes[T]>(
  type: T,
  initialValue: J,
  extras?: Partial<UnitConfig<T>>
): UniformUnit<T, J> => {
  const uniform = { value: initialValue }

  /* Create the actual unit that represents the uniform. */
  const unit = Unit(type, undefined, {
    name: `Uniform (${type})`,
    ...extras,
    uniform
  })

  /* Return the unit with some API bits mixed in. */
  return injectAPI(unit, () => ({
    set value(v: J) {
      uniform.value = v
    },

    get value(): J {
      return uniform.value
    }
  }))
}

/**
 * Provides a uniform unit holding a representation of time. The time value
 * stored is not an absolute time, so multiple instances of this unit will not
 * be synchronized. If you require synchronization, please either reuse the
 * same instance of this unit, or use `GlobalTime` instead.
 *
 * @param initial The initial time value to start with. (Default: 0)
 */
export const Time = (initial: number = 0) => {
  const uniform = UniformUnit("float", initial, {
    name: "Time Uniform",

    update: (dt) => {
      uniform.value += dt
    }
  })

  return uniform
}

/**
 * A global time uniform unit that can be safely used across multiple shaders,
 * wherever synchronization is required,
 * and as a default value for `time` inputs of other unit implementations, to prevent
 * shaders from being spammed by multiple uniforms all holding different
 * representations of time.
 */
export const GlobalTime = Time()

export const Resolution = UniformUnit("vec2", new Vector2(0, 0), {
  name: "Current Render Resolution",

  update: (dt, camera, scene, gl) => {
    Resolution.value.x = gl.domElement.width
    Resolution.value.y = gl.domElement.height
  }
})

export const CameraNear = UniformUnit("float", 0 as number, {
  name: "Camera Near Plane",

  update: (_, camera) => {
    if (camera instanceof PerspectiveCamera) {
      CameraNear.value = camera.near
    }
  }
})

export const CameraFar = UniformUnit("float", 0 as number, {
  name: "Camera Far Plane",

  update: (_, camera) => {
    if (camera instanceof PerspectiveCamera) {
      CameraFar.value = camera.far
    }
  }
})

export const ScreenUV = Vec2($`${FragmentCoordinate} / ${Resolution}`, {
  name: "Screen UV"
})
