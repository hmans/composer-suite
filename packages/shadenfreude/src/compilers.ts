import { Vector2 } from "three"
import { isExpression } from "./expressions"
import { glslRepresentation } from "./glslRepresentation"
import {
  assignment,
  block,
  concatenate,
  flatten,
  identifier,
  isSnippet,
  Parts,
  sluggify,
  Snippet,
  statement
} from "./lib/concatenator3000"
import idGenerator from "./lib/idGenerator"
import { isVariable, Variable } from "./variables"

export type ProgramType = "vertex" | "fragment"

const variableBeginComment = (v: Variable) =>
  `/*** BEGIN: ${v._config.title} (${v._config.id}) ***/`

const variableEndComment = (v: Variable) =>
  `/*** END: ${v._config.title} (${v._config.id}) ***/\n`

/**
 * Traverses the specified variables and returns a list of all objects that can
 * be dependencies of something else (variables, expressions, snippets, etc.)
 *
 * @param sources
 * @returns
 */
const getDependencies = (...sources: any[]): any[] =>
  sources
    .flat(Infinity)
    .map((s) =>
      isVariable(s)
        ? s
        : isExpression(s)
        ? [s.values, getDependencies(...s.values)]
        : isSnippet(s)
        ? s
        : undefined
    )
    .flat(Infinity)
    .filter((d) => !!d)

const compileSnippet = (
  s: Snippet,
  program: "vertex" | "fragment",
  state: ReturnType<typeof compilerState>
) => {
  if (!state.isFresh(s)) return

  if (isExpression(s.chunk))
    s.chunk.values.forEach((v) => {
      isVariable(v) && compileVariable(v, program, state)
      isSnippet(v) && compileSnippet(v, program, state)
    })

  state.header.push(`/*** SNIPPET: ${s.name} ***/`, s.chunk)
}

export const compileVariable = (
  v: Variable,
  program: ProgramType,
  state = compilerState()
) => {
  if (!state.isFresh(v)) return []
  if (v._config.only && v._config.only !== program) return []

  /* Build a list of dependencies from the various places that can have them: */
  const dependencies = getDependencies(
    v.value,
    v._config.fragmentHeader,
    v._config.fragmentBody,
    v._config.vertexHeader,
    v._config.vertexBody
  )

  /* Render variable and snippet dependencies */
  dependencies.forEach((dep) => {
    isVariable(dep) && compileVariable(dep, program, state)
    isSnippet(dep) && compileSnippet(dep, program, state)
  })

  /* Prepare this variable */
  v._config.id = state.nextId()
  v._config.name = identifier(v.type, sluggify(v._config.title), v._config.id)

  /* HEADER */
  const header = flatten(
    /* If this variable is configured to use a varying, declare it */
    v._config.varying && `varying ${v.type} v_${v._config.name};`,

    /* Render the actual header chuink */
    v._config[`${program}Header`]
  )

  state.header.push(
    /* Render header chunk */
    header.length && [variableBeginComment(v), header, variableEndComment(v)]
  )

  /* BODY */
  state.body.push(
    variableBeginComment(v),

    /* Declare the variable */
    statement(v.type, v._config.name),

    block(
      /* Make local value variable available */
      v._config.varying && program === "fragment"
        ? statement(v.type, "value", "=", `v_${v._config.name}`)
        : statement(v.type, "value", "=", glslRepresentation(v.value)),

      /* The body chunk, if there is one */
      v._config[`${program}Body`],

      /* Assign local value variable back to global variable */
      assignment(v._config.name, "value"),

      /* If we're in vertex and have a varying, assign to it, too */
      v._config.varying &&
        program === "vertex" &&
        assignment(`v_${v._config.name}`, "value")
    ),

    variableEndComment(v)
  )
}

/**  Compile a program from the variable */
export const compileProgram = (v: Variable, program: ProgramType) => {
  const state = compilerState()
  compileVariable(v, program, state)

  return concatenate(state.header, "void main()", block(state.body))
}

export const compileShader = (root: Variable) => {
  const vertexShader = compileProgram(root, "vertex")
  const fragmentShader = compileProgram(root, "fragment")

  const uniforms = {
    u_time: { value: 0 },
    u_resolution: { value: new Vector2() }
  }

  const update = (dt: number) => {
    uniforms.u_time.value += dt
    uniforms.u_resolution.value.set(window.innerWidth, window.innerHeight)
  }

  return [{ vertexShader, fragmentShader, uniforms }, update] as const
}

const compilerState = () => {
  const seen = new Set<Variable | Snippet>()
  const header = [] as Parts
  const body = [] as Parts
  const nextId = idGenerator()

  return {
    isFresh: (v: Variable | Snippet) =>
      seen.has(v) ? false : seen.add(v) && true,

    header,
    body,
    nextId
  }
}
