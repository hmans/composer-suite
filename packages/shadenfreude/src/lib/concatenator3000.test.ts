import { compileShader } from "../compilers"
import { code } from "../expressions"
import { Float } from "../variables"
import { concatenate, flatten, snippet } from "./concatenator3000"

describe("flatten", () => {
  it("doesn't modify a boring flat list of values", () => {
    const f = flatten(1, 2, 3)
    expect(f).toEqual([1, 2, 3])
  })

  it("filters blank parts", () => {
    const f = flatten(false, 1, "", 2, undefined, 3, null, 4)
    expect(f).toEqual([1, 2, 3, 4])
  })

  it("flattens nested arrays", () => {
    const f = flatten(1, 2, [3, [4, 5], 6], 7)
    expect(f).toEqual([1, 2, 3, 4, 5, 6, 7])
  })
})

describe("concatenate", () => {
  it("compiles the given parts into a string", () => {
    expect(concatenate("a", "b", "c")).toEqual("a\nb\nc")
    expect(concatenate("a", null, "c")).toEqual("a\nc")
    expect(concatenate("a", ["b"], "c")).toEqual("a\nb\nc")
  })
})

describe("snippet", () => {
  it("creates a Snippet with a unique name", () => {
    const s = snippet(() => "/* code */")
    expect(s.name).toEqual("snippet_a996254afd")
  })

  it("will generate the same snippet IDs for the same contents", () => {
    let code = "/* code */"
    const s1 = snippet(() => code)
    const s2 = snippet(() => code)
    expect(s1.name).toEqual("snippet_a996254afd")
    expect(s2.name).toEqual("snippet_a996254afd")
  })

  it("creates a snippet with a rendered chunk", () => {
    const s = snippet((name) => `/* hi from ${name} */`)
    expect(s).toMatchSnapshot()
  })

  it("will only render once per program", () => {
    const add = snippet(
      (name) => `float ${name}(float a, float b) { return a + b; }`
    )

    const f = Float(code`${add}(1, ${add}(2, 3))`)
    const [shader] = compileShader(f)

    expect(shader.vertexShader).toMatchInlineSnapshot(`
      "/*** SNIPPET: snippet_7dd01b876a ***/
      float snippet_7dd01b876a(float a, float b) { return a + b; }
      void main()
      {
        /*** BEGIN: anon (1) ***/
        float float_anon_1;
        {
          float value = snippet_7dd01b876a(1, snippet_7dd01b876a(2, 3));
          float_anon_1 = value;
        }
        /*** END: anon (1) ***/

      }"
    `)
  })

  it("allows a snippet to have dependencies to other snippets", () => {
    const mul = snippet(
      (name) => `float ${name}(float a, float b) { return a * b; }`
    )

    const add = snippet(
      (name) =>
        code`float ${name}(float a, float b) { return a + ${mul}(a, b); }`
    )

    const f = Float(code`${add}(1, ${add}(2, 3))`)
    const [shader] = compileShader(f)

    expect(shader.vertexShader).toMatchInlineSnapshot(`
      "/*** SNIPPET: snippet_3853dce07c ***/
      float snippet_3853dce07c(float a, float b) { return a * b; }
      /*** SNIPPET: snippet_98a183b64b ***/
      float snippet_98a183b64b(float a, float b) { return a + snippet_3853dce07c(a, b); }
      void main()
      {
        /*** BEGIN: anon (1) ***/
        float float_anon_1;
        {
          float value = snippet_98a183b64b(1, snippet_98a183b64b(2, 3));
          float_anon_1 = value;
        }
        /*** END: anon (1) ***/

      }"
    `)
  })
})
