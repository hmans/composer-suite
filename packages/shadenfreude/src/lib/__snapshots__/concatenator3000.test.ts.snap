// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snippet creates a snippet with a rendered chunk 1`] = `
Object {
  "_": "Snippet",
  "chunk": "/* hi from snippet_b117418551 */",
  "name": "snippet_b117418551",
}
`;
